opc-ua-sdk/sdk-client/src/main/java/org/eclipse/milo/opcua/sdk/client/subscriptions/OpcUaSubscription.java

```diff
@@ -200,12 +203,66 @@ public OpcUaSubscription(
         });
     }
 
+    @Override
+    public CompletableFuture<List<UaMonitoredItem>> transferMonitoredItems(
+        List<TransferMonitoredItemsRequest> itemsToTransfer) {
+        return notificationSemaphore.acquire().thenCompose(permit -> {
+            Map<TransferMonitoredItemsRequest, List<UaMonitoredItem>> monitoredItemsByRequest = new HashMap<>();
+            List<MonitoredItemModifyRequest> modifyRequests = newArrayList();
+
+            for (TransferMonitoredItemsRequest transferReq : itemsToTransfer) {
+                monitoredItemsByRequest.put(transferReq, newArrayList());
+
+                for (MonitoredItemTransferRequest monitoredItem : transferReq.getMonitoredItems()) {
+                    OpcUaMonitoredItem item = new OpcUaMonitoredItem(
+                        client,
+                        monitoredItem.getClientHandle(),
+                        monitoredItem.getItemToMonitor(),
+                        monitoredItem.getServerHandle(),
+                        new StatusCode(StatusCodes.Uncertain_InitialValue),
+                        monitoredItem.getRequestedParameters().getSamplingInterval(),
+                        monitoredItem.getRequestedParameters().getQueueSize(),
+                        monitoredItem.getRequestedParameters().getFilter(),
+                        monitoredItem.getMonitoringMode(),
+                        monitoredItem.getRequestedParameters().getFilter()
+                    );
+                    itemsByServerHandle.put(monitoredItem.getServerHandle(), item);
+                    itemsByClientHandle.put(monitoredItem.getClientHandle(), item);
+                    monitoredItemsByRequest.get(transferReq).add(item);
+                    modifyRequests.add(new MonitoredItemModifyRequest(
+                        monitoredItem.getServerHandle(),
+                        monitoredItem.getRequestedParameters()));
+                }
+            }
+            CompletableFuture<List<StatusCode>> future =
+                this.modifyMonitoredItems(TimestampsToReturn.Neither, modifyRequests);
+
+            return future.thenApply(modifyResponse -> {
+                try {
+                    monitoredItemsByRequest.forEach((req, items) -> {
+                        if (req.getItemTransferCallback().isPresent()) {
+                            ItemTransferredCallback callback = req.getItemTransferCallback().get();
+                            for (UaMonitoredItem item : items) {
+                                int clientHandle = item.getClientHandle().intValue();
+                                callback.onItemTransferred(client.getSerializationContext(), item, clientHandle);
+                            }
+                        }
+                    });
+                } finally {
+                    permit.release();
```

Comment:
Need to figure out a safer way to ensure `permit.release()` is called. If `modifyMonitoredItems()` call fails then the whole `thenApply` block never even runs.

Reply:
Something like this:
```java
CompletableFuture<List<UaMonitoredItem>> modifyFuture = new CompletableFuture<>();

modifyMonitoredItems(TimestampsToReturn.Neither, modifyRequests).thenApply(modifyResponse -> {
    monitoredItemsByRequest.forEach((req, items) -> {
        if (req.getItemTransferCallback().isPresent()) {
            ItemTransferredCallback callback = req.getItemTransferCallback().get();
            for (UaMonitoredItem item : items) {
                int clientHandle = item.getClientHandle().intValue();
                callback.onItemTransferred(client.getSerializationContext(), item, clientHandle);
            }
        }
    });

    return monitoredItemsByRequest.values()
        .stream()
        .flatMap(Collection::stream)
        .collect(toList());
});

return modifyFuture.whenComplete((r, ex) -> permit.release());
```

Reply:
Something like this maybe:
```java
return modifyMonitoredItems(TimestampsToReturn.Neither, modifyRequests)
    .thenApply(modifyResponse -> {
        monitoredItemsByRequest.forEach((req, items) -> {
            if (req.getItemTransferCallback().isPresent()) {
                ItemTransferredCallback callback = req.getItemTransferCallback().get();
                for (UaMonitoredItem item : items) {
                    int clientHandle = item.getClientHandle().intValue();
                    callback.onItemTransferred(client.getSerializationContext(), item, clientHandle);
                }
            }
        });

        return monitoredItemsByRequest.values()
            .stream()
            .flatMap(Collection::stream)
            .collect(toList());
    })
    .whenComplete((i, ex) -> permit.release());
```

Reply:
This seems reasonable, I'll test it with the other changes mentioned in this review.